src/treelearner/feature_histogram.hpp

```diff
@@ -240,16 +276,31 @@ class FeatureHistogram {
     }
 
     if (is_splittable_) {
-      output->left_output = CalculateSplittedLeafOutput(best_sum_left_gradient, best_sum_left_hessian,
-                                                        meta_->config->lambda_l1, l2, meta_->config->max_delta_step,
-                                                        min_constraint, max_constraint);
+      double current_min_constraint_left = -std::numeric_limits<double>::max();
+      double current_max_constraint_left = std::numeric_limits<double>::max();
+      double current_min_constraint_right = -std::numeric_limits<double>::max();
+      double current_max_constraint_right = std::numeric_limits<double>::max();
+      if (constraints != nullptr) {
+        current_min_constraint_left = constraints->CurrentMinConstraintLeft();
+        current_max_constraint_left = constraints->CurrentMaxConstraintLeft();
+        current_min_constraint_right = constraints->CurrentMinConstraintRight();
+        current_max_constraint_right = constraints->CurrentMaxConstraintRight();
+      }
```

Comment:
could we avoid these expansions?  maybe have two structures for two leaves, e.g. 
```constraints->left->min```, ```constraints->left->max``` ,```constraints->right->min```, ```constraints->right->max```.
Then pass `constraints->left` and `constraints->right` to CalculateSplittedLeafOutput, respectively.